haystack/nodes/prompt/prompt_node.py

```diff
@@ -897,21 +897,25 @@ def run(
         :param meta: The meta to be used for the prompt. Usually not used.
         :param invocation_context: The invocation context to be used for the prompt.
         """
+        # The invocation context overwrites locals(), so if for example invocation_context contains a
+        # modified list of Documents under the `documents` key, such list is used.
+        invocation_context = {**locals(), **(invocation_context or {})}
+        invocation_context.pop("self")
 
-        # invocation_context is a dictionary that is passed from a pipeline node to a pipeline node and can be used
-        # to pass results from a pipeline node to any other downstream pipeline node.
-        invocation_context = invocation_context or {}
+        for doc in invocation_context.get("documents", []):
+            if not isinstance(doc, str) and not isinstance(doc.content, str):
+                raise ValueError("PromptNode only accepts text documents.")
 
-        results = self(
-            query=query,
-            labels=labels,
-            documents=[doc.content for doc in documents if isinstance(doc.content, str)] if documents else [],
-            **invocation_context,
-        )
+        invocation_context["documents"] = [
+            doc.content if isinstance(doc, Document) else doc for doc in invocation_context.get("documents", [])
+        ]
+
+        results = self(**invocation_context)
 
         if self.output_variable:
             invocation_context[self.output_variable] = results
-        return {"results": results, "invocation_context": invocation_context}, "output_1"
+
+        return {"results": results, "meta": meta, "invocation_context": invocation_context}, "output_1"
```

Comment:
I think we don't need to include meta here. Everything related to PromptNode is in invocation_context.